presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/OptimizeMixedDistinctAggregations.java

```diff
@@ -0,0 +1,527 @@
+                aggregateInfo,
+                distinctSymbol,
+                duplicatedDistinctSymbol,
+                groupByKeys,
+                groupIdNode,
+                markDistinctNode,
+                symbolAllocator,
+                idAllocator);
+
+        // 3. Add new project node that adds IF expressions with original symbol mapped
+        return createProjectNodeInfo(
+                nonDistinctAggregationInfo,
+                aggregateInfo.getMask(),
+                distinctSymbol,
+                groupSymbol,
+                groupBySymbols,
+                symbolAllocator,
+                idAllocator);
+    }
+
+    // Returns false if either the mask symbol or any of the symbols in aggregations are not comparable
+    private static boolean checkAllEquatableTypes(AggregateInfo aggregateInfo, SymbolAllocator symbolAllocator)
+    {
+        for (Symbol symbol : aggregateInfo.getOriginalNonDistinctAggregateArgs()) {
+            Type type = symbolAllocator.getTypes().get(symbol);
+            if (!type.isComparable()) {
+                return false;
+            }
+        }
+
+        return symbolAllocator.getTypes().get(aggregateInfo.getMask()).isComparable();
+    }
+
+    private static GroupIdNode createGroupIdNode(
+            List<Symbol> groupBySymbols,
+            List<Symbol> nonDistinctAggregateSymbols,
+            Symbol distinctSymbol,
+            Optional<Symbol> duplicatedDistinctSymbol,
+            Symbol groupSymbol,
+            Set<Symbol> allSymbols,
+            PlanNode source,
+            PlanNodeIdAllocator idAllocator)
+    {
+        List<List<Symbol>> groups = new ArrayList<>();
+        // g0 = {group-by symbols + allNonDistinctAggregateSymbols}
+        // g1 = {group-by symbols + Distinct Symbol}
+        // symbols present in Group_i will be set, rest will be Null
+
+        //g0
+        Set<Symbol> group0 = new HashSet<>();
+        group0.addAll(groupBySymbols);
+        group0.addAll(nonDistinctAggregateSymbols);
+        groups.add(ImmutableList.copyOf(group0));
+
+        // g1
+        Set<Symbol> group1 = new HashSet<>();
+        group1.addAll(groupBySymbols);
+        group1.add(distinctSymbol);
+        groups.add(ImmutableList.copyOf(group1));
+
+        return new GroupIdNode(
+                idAllocator.getNextId(),
+                source,
+                groups,
+                allSymbols.stream().distinct().collect(Collectors.toMap(
+                        symbol -> symbol,
+                        symbol -> (duplicatedDistinctSymbol.isPresent() && symbol.equals(duplicatedDistinctSymbol.get()) ? distinctSymbol : symbol))),
+                ImmutableMap.of(),
+                groupSymbol);
+    }
+
+    /*
+     * This method returns a new Aggregation node which has aggregations on non-distinct symbols from original plan. Generates
+     *      SELECT a1, a2,..., an, F1(b1) as f1, F2(b2) as f2,...., Fm(bm) as fm, c, group
```

Comment:
maybe express this in terms of the plan node since there is no longer a sql description of the rewrite at the top. same for similar comments elsewhere.